source/core_mqtt.c

```diff
@@ -1532,56 +1758,318 @@ static MQTTStatus_t validateSubscribeUnsubscribeParams( const MQTTContext_t * pC
+
+static MQTTStatus_t sendPublishWithoutCopy( MQTTContext_t * pContext,
+                                            const MQTTPublishInfo_t * pPublishInfo,
+                                            const uint8_t * pMqttHeader,
+                                            size_t headerSize,
+                                            uint16_t packetId )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    uint8_t serializedPacketID[ 2 ];
+    TransportOutVector_t pIoVector[ 4 ];
+    size_t ioVectorLength;
+    size_t totalMessageLength;
+    const size_t packetIDLength = 2U;
+
+    /* The header is sent first. */
+    pIoVector[ 0U ].iov_base = pMqttHeader;
+    pIoVector[ 0U ].iov_len = headerSize;
+    totalMessageLength = headerSize;
+
+    /* Then the topic name has to be sent. */
+    pIoVector[ 1U ].iov_base = pPublishInfo->pTopicName;
+    pIoVector[ 1U ].iov_len = pPublishInfo->topicNameLength;
+    totalMessageLength += pPublishInfo->topicNameLength;
+
+    /* The next field's index should be 2 as the first two fields
+     * have been filled in. */
+    ioVectorLength = 2U;
+
+    if( pPublishInfo->qos > MQTTQoS0 )
+    {
+        /* Encode the packet ID. */
+        serializedPacketID[ 0 ] = ( ( uint8_t ) ( ( packetId ) >> 8 ) );
+        serializedPacketID[ 1 ] = ( ( uint8_t ) ( ( packetId ) & 0x00ffU ) );
+
+        pIoVector[ ioVectorLength ].iov_base = serializedPacketID;
+        pIoVector[ ioVectorLength ].iov_len = packetIDLength;
+
+        ioVectorLength++;
+        totalMessageLength += packetIDLength;
+    }
+
+    /* Publish packets are allowed to contain no payload. */
+    if( pPublishInfo->payloadLength > 0U )
+    {
+        pIoVector[ ioVectorLength ].iov_base = pPublishInfo->pPayload;
+        pIoVector[ ioVectorLength ].iov_len = pPublishInfo->payloadLength;
+
+        ioVectorLength++;
+        totalMessageLength += pPublishInfo->payloadLength;
+    }
+
+    if( sendMessageVector( pContext, pIoVector, ioVectorLength ) != ( int32_t ) totalMessageLength )
     {
-        LogError( ( "Transport send failed for PUBLISH header." ) );
         status = MQTTSendFailed;
     }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static void addWillAndConnectInfo( const MQTTConnectInfo_t * pConnectInfo,
+                                   const MQTTPublishInfo_t * pWillInfo,
+                                   size_t * pTotalMessageLength,
+                                   TransportOutVector_t * iterator,
+                                   uint8_t serializedTopicLength[ 2 ],
+                                   uint8_t serializedPayloadLength[ 2 ],
+                                   uint8_t serializedUsernameLength[ 2 ],
+                                   uint8_t serializedPasswordLength[ 2 ] )
+{
+    if( pWillInfo != NULL )
+    {
+        /* Serialize the topic. */
+        iterator = addEncodedStringToVector( serializedTopicLength,
```

Comment:
I'm not sure that this improves clarity really. Wouldn't it be simpler just to fill the array?